File: installers/python/src/blender_mobile_3d_installer/download.py
"""Secure artifact download: HTTPS-or-loopback, bounded size, atomic rename."""

from __future__ import annotations

import hashlib
import tempfile
import urllib.request
from pathlib import Path
from urllib.error import URLError
from urllib.parse import urlparse

from blender_mobile_3d_installer.exit_codes import (
    CHECKSUM_MISMATCH,
    DOWNLOAD_FAILURE,
    InstallerError,
)

DOWNLOAD_TIMEOUT_SECONDS = 60
DOWNLOAD_CHUNK_BYTES = 1024 * 1024
_LOOPBACK_HOSTS = {"localhost", "127.0.0.1", "::1"}


def _require_https_or_loopback(url: str) -> None:
    parsed = urlparse(url)
    if parsed.scheme == "https":
        return
    if parsed.scheme == "http" and parsed.hostname in _LOOPBACK_HOSTS:
        return
    raise InstallerError(
        DOWNLOAD_FAILURE,
        f"Refusing non-HTTPS download URL (loopback exempt for testing only): {url}",
    )


def download_artifact(
    url: str,
    dest_path: Path,
    expected_sha256: str,
    max_size_bytes: int,
) -> Path:
    """Download ``url`` to ``dest_path``, verifying size and checksum.

    Writes to a temporary file in the destination's own directory and
    performs an atomic rename only after the checksum matches. Any failure
    removes the partial file and raises InstallerError.
    """
    _require_https_or_loopback(url)
    dest_path.parent.mkdir(parents=True, exist_ok=True)

    digest = hashlib.sha256()
    total = 0
    tmp_fd, tmp_name = tempfile.mkstemp(
        dir=str(dest_path.parent), prefix=".download-", suffix=".tmp"
    )
    tmp_path = Path(tmp_name)
    try:
        try:
            # Scheme already restricted to https (or loopback http for tests)
            # by _require_https_or_loopback above.
            request = urllib.request.Request(  # noqa: S310
                url, headers={"User-Agent": "blender-mobile-3d-installer"}
            )
            with (
                urllib.request.urlopen(request, timeout=DOWNLOAD_TIMEOUT_SECONDS) as response,  # noqa: S310
                open(tmp_fd, "wb") as tmp_file,
            ):
                while True:
                    chunk = response.read(DOWNLOAD_CHUNK_BYTES)
                    if not chunk:
                        break
                    total += len(chunk)
                    if total > max_size_bytes:
                        raise InstallerError(
                            DOWNLOAD_FAILURE,
                            f"Download from {url} exceeded the {max_size_bytes}-byte limit.",
                        )
                    digest.update(chunk)
                    tmp_file.write(chunk)
        except (URLError, TimeoutError, OSError) as exc:
            raise InstallerError(DOWNLOAD_FAILURE, f"Download failed for {url}: {exc}") from exc

        if total == 0:
            raise InstallerError(DOWNLOAD_FAILURE, f"Download from {url} was empty.")

        actual = digest.hexdigest()
        if actual != expected_sha256:
            raise InstallerError(
                CHECKSUM_MISMATCH,
                f"Checksum mismatch for {url}: expected {expected_sha256}, got {actual}.",
            )

        tmp_path.replace(dest_path)
        return dest_path
    finally:
        tmp_path.unlink(missing_ok=True)

File: installers/npm/src/download.js
/** Secure artifact download: HTTPS-or-loopback, bounded size, atomic rename. */

import { createHash } from "node:crypto";
import { createWriteStream } from "node:fs";
import { mkdir, mkdtemp, rename, rm, stat } from "node:fs/promises";
import path from "node:path";
import { Transform } from "node:stream";
import { pipeline } from "node:stream/promises";

import { CHECKSUM_MISMATCH, DOWNLOAD_FAILURE, InstallerError } from "./exitCodes.js";

export const DOWNLOAD_TIMEOUT_MS = 60_000;
const LOOPBACK_HOSTS = new Set(["localhost", "127.0.0.1", "::1"]);

function requireHttpsOrLoopback(url) {
  const parsed = new URL(url);
  if (parsed.protocol === "https:") return;
  if (parsed.protocol === "http:" && LOOPBACK_HOSTS.has(parsed.hostname)) return;
  throw new InstallerError(
    DOWNLOAD_FAILURE,
    `Refusing non-HTTPS download URL (loopback exempt for testing only): ${url}`,
  );
}

/**
 * Download `url` to `destPath`, verifying size and checksum. Writes to a
 * temporary file in the destination's own directory and performs an atomic
 * rename only after the checksum matches. Any failure removes the partial
 * file and throws InstallerError.
 */
export async function downloadArtifact(url, destPath, expectedSha256, maxSizeBytes) {
  requireHttpsOrLoopback(url);
  await mkdir(path.dirname(destPath), { recursive: true });

  const tmpDir = await mkdtemp(path.join(path.dirname(destPath), ".download-"));
  const tmpPath = path.join(tmpDir, "artifact.tmp");

  try {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), DOWNLOAD_TIMEOUT_MS);
    let response;
    try {
      response = await fetch(url, {
        signal: controller.signal,
        headers: { "User-Agent": "blender-mobile-3d-installer" },
      });
    } catch (err) {
      throw new InstallerError(DOWNLOAD_FAILURE, `Download failed for ${url}: ${err.message}`);
    } finally {
      clearTimeout(timer);
    }

    if (!response.ok || !response.body) {
      throw new InstallerError(DOWNLOAD_FAILURE, `Download failed for ${url}: HTTP ${response.status}`);
    }

    const hash = createHash("sha256");
    let total = 0;
    const sizeGuard = new Transform({
      transform(chunk, _enc, callback) {
        total += chunk.length;
        if (total > maxSizeBytes) {
          callback(new Error(`exceeded the ${maxSizeBytes}-byte limit`));
          return;
        }
        hash.update(chunk);
        callback(null, chunk);
      },
    });

    try {
      await pipeline(response.body, sizeGuard, createWriteStream(tmpPath));
    } catch (err) {
      throw new InstallerError(DOWNLOAD_FAILURE, `Download failed for ${url}: ${err.message}`);
    }

    if (total === 0) {
      throw new InstallerError(DOWNLOAD_FAILURE, `Download from ${url} was empty.`);
    }

    const actual = hash.digest("hex");
    if (actual !== expectedSha256) {
      throw new InstallerError(
        CHECKSUM_MISMATCH,
        `Checksum mismatch for ${url}: expected ${expectedSha256}, got ${actual}.`,
      );
    }

    await rename(tmpPath, destPath);
    return destPath;
  } finally {
    await rm(tmpDir, { recursive: true, force: true });
  }
}

export async function fileExists(candidatePath) {
  try {
    await stat(candidatePath);
    return true;
  } catch {
    return false;
  }
}
